src/shared/utilities/timeoutUtils.ts

```diff
@@ -168,29 +168,29 @@ export async function waitUntil<T>(
  *
  * @returns A Promise that returns if successful, or rejects when the Timeout was cancelled or expired.
  */
-export function waitTimeout<T>(
+export function waitTimeout<T, R = void, B extends boolean = true>(
     promise: Promise<T> | (() => Promise<T>),
-    timeout: Timeout,
+    timeout: Timeout, // TODO: potentially type 'completed' timers differently from active ones
     opt: {
-        allowUndefined?: boolean
-        onExpire?: () => T | undefined
-        onCancel?: () => T | undefined
+        allowUndefined?: B
+        onExpire?: () => R
+        onCancel?: () => R
         completeTimeout?: boolean
     } = {}
-): Promise<T | undefined> {
+): Promise<T | R | (true extends typeof opt.allowUndefined ? undefined : never)> {
     if (typeof promise === 'function') {
         promise = promise()
     }
 
     return Promise.race([promise, timeout.timer])
         .then(obj => {
-            if (opt.allowUndefined === false && obj === undefined) {
-                throw new Error(TIMEOUT_UNEXPECTED_RESOLVE)
-            }
             if (obj !== undefined) {
                 return obj
             }
-            return undefined
+            if ((opt.allowUndefined ?? true) !== true) {
+                throw new Error(TIMEOUT_UNEXPECTED_RESOLVE)
+            }
+            return undefined as true extends typeof opt.allowUndefined ? undefined : never
```

Comment:
what's going on here , is the readability loss worth whatever we  gains vs `return undefined` ?

Reply:
TypeScript can't keep track of typing in some niche flow control scenarios so this manually annotates the type corresponding to the flow. This keeps the implementation self-consistent, but it does look pretty ugly (especially if you're not used to `extends` being using for conditional types, GitHub doesn't help either). Since even this annotation doesn't really capture the full meaning, it's best to just cast as `any` for now, and let unit tests validate behavior instead.